tikv-client/src/main/java/com/pingcap/tikv/region/RegionManager.java

```diff
@@ -183,121 +261,20 @@ public void invalidateRegion(TiRegion region) {
-        }
-        TiRegion oldRegion = regionCache.get(getEncodedKey(region.getStartKey()));
-        if (oldRegion == null || !expected.getMeta().equals(oldRegion.getMeta())) {
-          return false;
-        } else {
-          regionCache.remove(makeRange(oldRegion.getStartKey(), oldRegion.getEndKey()));
-          regionCache.put(makeRange(region.getStartKey(), region.getEndKey()), region);
-          return true;
-        }
-      } catch (Exception ignore) {
-        return false;
-      }
-    }
-
-    private synchronized TiRegion getRegionFromCache(Key key) {
-      return regionCache.get(key);
-    }
-
-    private synchronized void invalidateRange(ByteString startKey, ByteString endKey) {
-      regionCache.remove(makeRange(startKey, endKey));
-    }
-
-    /** Removes region associated with regionId from regionCache. */
-    public synchronized void invalidateRegion(TiRegion region) {
+  private TiRegion createRegion(Metapb.Region region, Metapb.Peer leader, BackOffer backOffer) {
+    List<Metapb.Peer> peers = new ArrayList<>();
+    List<TiStore> stores = new ArrayList<>();
+    for (Metapb.Peer peer : region.getPeersList()) {
       try {
-        if (logger.isDebugEnabled()) {
-          logger.debug(String.format("invalidateRegion ID[%s]", region.getId()));
-        }
-        TiRegion oldRegion = regionCache.get(getEncodedKey(region.getStartKey()));
-        if (oldRegion != null && oldRegion.equals(region)) {
-          regionCache.remove(makeRange(region.getStartKey(), region.getEndKey()));
-        }
-      } catch (Exception ignore) {
-      }
-    }
-
-    public synchronized void invalidateAllRegionForStore(long storeId) {
-      List<TiRegion> regionToRemove = new ArrayList<>();
-      for (TiRegion r : regionCache.asMapOfRanges().values()) {
-        if (r.getLeader().getStoreId() == storeId) {
-          if (logger.isDebugEnabled()) {
-            logger.debug(String.format("invalidateAllRegionForStore Region[%s]", r));
-          }
-          regionToRemove.add(r);
-        }
-      }
-
-      // remove region
-      for (TiRegion r : regionToRemove) {
-        regionCache.remove(makeRange(r.getStartKey(), r.getEndKey()));
-      }
-    }
-
-    public synchronized void invalidateStore(long storeId) {
-      storeCache.remove(storeId);
-    }
-
-    public synchronized Store getStoreById(long id, BackOffer backOffer) {
-      try {
-        Store store = storeCache.get(id);
-        if (store == null) {
-          store = pdClient.getStore(backOffer, id);
-        }
-        if (store.getState().equals(StoreState.Tombstone)) {
-          return null;
-        }
-        storeCache.put(id, store);
-        return store;
+        stores.add(getStoreById(peer.getStoreId(), backOffer));
+        peers.add(peer);
       } catch (Exception e) {
-        throw new GrpcException(e);
+        logger.warn("Store {} not found: {}", peer.getStoreId(), e.toString());
```

Comment:
Why change the behavior?

Reply:
This is because getStoreById needs to throw exceptions before, but now our getStoreById will still throw exceptions, and createRegion needs to ignore exceptions generated by calling getStoreById.